core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -273,6 +273,9 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
 
         val brokerInfo = createBrokerInfo
         val brokerEpoch = zkClient.registerBroker(brokerInfo)
+        if (config.preferredController) {
+          zkClient.registerPreferredControllerId(brokerInfo.broker.id)
```

Comment:
Why do we need to register preferred controller id here given that we will do it in `KafkaController::processRegisterBrokerAndReelect ` anyway during controller startup? Again, we can avoid doing this by embedding the preferred controller flag in brokerInfo.

Reply:
processRegisterBrokerAndReelect is only called after zk session initialization. For startup, processRegisterBrokerAndReelect will not be called so we need a separated registration.